src/ZenohUTransport.cpp

```diff
@@ -10,3 +10,356 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "up-transport-zenoh-cpp/ZenohUTransport.h"
+
+#include <up-cpp/datamodel/serializer/UUri.h>
+#include <up-cpp/datamodel/serializer/Uuid.h>
+#include <up-cpp/datamodel/validator/UMessage.h>
+#include <up-cpp/datamodel/validator/UUri.h>
+
+#include <stdexcept>
+
+namespace uprotocol::transport {
+
+const char UATTRIBUTE_VERSION = 1;
+
+const uint32_t WILDCARD_ENTITY_ID = 0x0000FFFF;
+const uint32_t WILDCARD_ENTITY_VERSION = 0x000000FF;
+const uint32_t WILDCARD_RESOURCE_ID = 0x0000FFFF;
+
+using namespace zenoh;
+using namespace uprotocol::v1;
+using namespace uprotocol::datamodel;
+
+UStatus ZenohUTransport::uError(UCode code, std::string_view message) {
+	UStatus status;
+	status.set_code(code);
+	status.set_message(std::string(message));
+	return status;
+}
+
+std::string ZenohUTransport::toZenohKeyString(
+    const std::string& default_authority_name, const UUri& source,
+    const std::optional<v1::UUri>& sink) {
+	std::ostringstream zenoh_key;
+
+	auto writeUUri = [&](const v1::UUri& uuri) {
+		zenoh_key << "/";
+
+		// authority_name
+		if (uuri.authority_name().empty()) {
+			zenoh_key << default_authority_name;
+		} else {
+			zenoh_key << uuri.authority_name();
+		}
+		zenoh_key << "/";
+
+		// ue_id
+		if (uuri.ue_id() == WILDCARD_ENTITY_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_id();
+		}
+		zenoh_key << "/";
+
+		// ue_version_major
+		if (uuri.ue_version_major() == WILDCARD_ENTITY_VERSION) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_version_major();
+		}
+		zenoh_key << "/";
+
+		// resource_id
+		if (uuri.resource_id() == WILDCARD_RESOURCE_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.resource_id();
+		}
+	};
+
+	zenoh_key << "up";
+	zenoh_key << std::uppercase << std::hex;
+
+	writeUUri(source);
+
+	if (sink.has_value()) {
+		writeUUri(*sink);
+	} else {
+		zenoh_key << "/{}/{}/{}/{}";
+	}
+	return zenoh_key.str();
+}
+
+std::vector<std::pair<std::string, std::string>>
+ZenohUTransport::uattributesToAttachment(const UAttributes& attributes) {
+	std::vector<std::pair<std::string, std::string>> res;
+
+	std::string version(&UATTRIBUTE_VERSION, 1);
+
+	std::string data;
+	attributes.SerializeToString(&data);
+
+	res.push_back(std::make_pair("", version));
```

Comment:
Should this be emplace_back instead of push_back?

Reply:
Agree